File: src/types/autosave.ts
// Auto-Save Management System - TypeScript Type Definitions

// ============================================================================
// Core Auto-Save Types
// ============================================================================

export interface AutoSave {
  id: number;
  serverId: number;
  fileName: string;
  filePath: string;
  fileSize: number;
  checksum?: string;
  createdAt: string;
  saveTimestamp?: string;
  isValid: boolean;
  isCorrupted: boolean;
  corruptionReason?: string;
  playerCount?: number;
  uptimeSeconds?: number;
  serverVersion?: string;
  modCount?: number;
  mapName?: string;
  isProtected: boolean;
  customLabel?: string;
  notes?: string;
  isFavorite: boolean;
  folderId?: number;
  createdBy?: string;
  indexedAt: string;
  updatedAt: string;
}

export interface AutoSaveRequest {
  serverId: number;
  fileName: string;
  filePath: string;
  fileSize: number;
  playerCount?: number;
  uptimeSeconds?: number;
  serverVersion?: string;
  modCount?: number;
  mapName?: string;
}

// ============================================================================
// Save Folder Types
// ============================================================================

export interface SaveFolder {
  id: number;
  name: string;
  description?: string;
  color: string;
  parentFolderId?: number;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
  children?: SaveFolder[];
  saveCount?: number;
}

export interface SaveFolderRequest {
  name: string;
  description?: string;
  color?: string;
  parentFolderId?: number;
}

// ============================================================================
// Save Restore Types
// ============================================================================

export interface SaveRestoreHistory {
  id: number;
  serverId: number;
  fromSaveId?: number;
  toSaveId: number;
  restoredAt: string;
  restoredBy?: string;
  restoreDurationSeconds?: number;
  restoreMethod: string; // manual, automatic, scheduled, emergency
  success: boolean;
  errorMessage?: string;
  notes?: string;
}

export interface RestoreRequest {
  serverId: number;
  saveId: number;
  createBackup: boolean;
  restoreMethod: string;
  skipRestart?: boolean;
}

export interface SaveBackupSnapshot {
  id: number;
  serverId: number;
  restoreHistoryId?: number;
  backupPath: string;
  backupSize: number;
  createdAt: string;
  isValid: boolean;
}

// ============================================================================
// Save Metadata Types
// ============================================================================

export interface SaveMetadata {
  id: number;
  autoSaveId: number;
  gameMode?: string;
  difficultyLevel?: string;
  maxPlayers?: number;
  currentPlayersList: string[];
  modsList: ModInfo[];
  serverSettings: Record<string, any>;
  worldStatistics: Record<string, any>;
  creaturesCount?: number;
  structuresCount?: number;
  itemsCount?: number;
  parsedAt: string;
}

export interface ModInfo {
  name: string;
  version?: string;
  modId?: string;
}

// ============================================================================
// Validation Types
// ============================================================================

export interface SaveValidationLog {
  id: number;
  autoSaveId: number;
  checkType: string; // integrity, corruption, completeness, compatibility
  checkStatus: string; // passed, failed, warning, skipped
  details: Record<string, string>;
  checkedAt: string;
}

export interface SaveValidationResult {
  isValid: boolean;
  isCorrupted: boolean;
  corruptionReason?: string;
  warnings: string[];
  errors: string[];
  fileSizeOk: boolean;
  checksumOk?: boolean;
  compatible: boolean;
}

// ============================================================================
// Restore Points Types
// ============================================================================

export interface RestorePoint {
  id: number;
  serverId: number;
  autoSaveId: number;
  pointName: string;
  description?: string;
  pointType: string; // manual, scheduled, critical, post-crash
  createdAt: string;
  isProtected: boolean;
  createdBy?: string;
}

export interface CreateRestorePointRequest {
  serverId: number;
  saveId: number;
  pointName: string;
  description?: string;
  pointType?: string;
}

// ============================================================================
// Save Comparison Types
// ============================================================================

export interface SaveComparison {
  saveId1: number;
  saveId2: number;
  sizeDifference: number;
  timestampDifferenceSeconds: number;
  playerCountDifference: number;
  uptimeDifferenceSeconds: number;
  modChanges: ModChanges;
  settingChanges: SettingChange[];
  hasCorruptionChange: boolean;
}

export interface ModChanges {
  addedMods: string[];
  removedMods: string[];
  updatedMods: ModUpdate[];
}

export interface ModUpdate {
  modName: string;
  oldVersion: string;
  newVersion: string;
}

export interface SettingChange {
  settingName: string;
  oldValue: string;
  newValue: string;
}

// ============================================================================
// Timeline Types
// ============================================================================

export interface TimelineEvent {
  id: number;
  serverId: number;
  eventType: string; // save_created, restored, protected, deleted, validated
  autoSaveId?: number;
  restoreHistoryId?: number;
  eventTime: string;
  description: string;
  metadata: Record<string, string>;
  importanceLevel: string; // low, normal, high, critical
  createdAt: string;
}

export interface TimelineResponse {
  events: TimelineEvent[];
  totalEvents: number;
  serverId: number;
}

// ============================================================================
// Archive and Compression Types
// ============================================================================

export interface SaveArchive {
  id: number;
  serverId: number;
  archivePath: string;
  archiveSize: number;
  saveIds: number[];
  createdAt: string;
  isVerified: boolean;
  lastAccessed?: string;
}

export interface CompressionRequest {
  serverId: number;
  saveIds: number[];
  compressionLevel?: number;
}

// ============================================================================
// Cloud Sync Types
// ============================================================================

export interface SaveCloudSync {
  id: number;
  autoSaveId: number;
  cloudPath?: string;
  syncStatus: string; // pending, syncing, synced, failed
  lastSyncAt?: string;
  syncError?: string;
  cloudVersion?: string;
}

// ============================================================================
// Preferences Types
// ============================================================================

export interface AutosavePreferences {
  id: number;
  serverId: number;
  autoIndexEnabled: boolean;
  autoValidateEnabled: boolean;
  autoCompressOldSaves: boolean;
  compressAfterDays: number;
  autoCleanupEnabled: boolean;
  cleanupAfterDays: number;
  keepMinimumSaves: number;
  createRestorePoints: boolean;
  restorePointFrequency: string; // hourly, daily, weekly, manual
  notifyOnRestore: boolean;
  notifyOnCorruption: boolean;
  indexMetadata: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface UpdatePreferencesRequest {
  serverId: number;
  autoIndexEnabled?: boolean;
  autoValidateEnabled?: boolean;
  autoCompressOldSaves?: boolean;
  compressAfterDays?: number;
  autoCleanupEnabled?: boolean;
  cleanupAfterDays?: number;
  keepMinimumSaves?: number;
  createRestorePoints?: boolean;
  restorePointFrequency?: string;
  notifyOnRestore?: boolean;
  notifyOnCorruption?: boolean;
  indexMetadata?: boolean;
}

// ============================================================================
// Statistics Types
// ============================================================================

export interface SaveStatistics {
  serverId: number;
  totalSaves: number;
  protectedSaves: number;
  corruptedSaves: number;
  totalStorageUsed: number;
  oldestSaveDate?: string;
  newestSaveDate?: string;
  averageSaveSize: number;
  totalRestores: number;
  lastRestoreDate?: string;
  favoriteCount: number;
}

export interface SaveHealthStatus {
  healthScore: number; // 0-100
  status: string; // excellent, good, fair, poor, critical
  issuesCount: number;
  warnings: string[];
  recommendations: string[];
}

// ============================================================================
// Bulk Operation Types
// ============================================================================

export interface BulkSaveOperation {
  operationType: string; // delete, protect, tag, compress, export
  saveIds: number[];
  operationData: Record<string, any>;
}

export interface BulkOperationResult {
  operationId: number;
  status: string; // pending, in-progress, completed, failed
  processedCount: number;
  errorCount: number;
  startedAt: string;
  completedAt?: string;
  errors: string[];
}

// ============================================================================
// Search and Filter Types
// ============================================================================

export interface SaveSearchFilter {
  searchQuery?: string;
  serverId?: number;
  dateRange?: DateRange;
  status?: string[]; // valid, corrupted, protected, favorite
  mapNames?: string[];
  minSize?: number;
  maxSize?: number;
  playerCountRange?: [number, number];
}

export interface DateRange {
  start: string;
  end: string;
}

export interface SaveSortOptions {
  sortBy: string; // date, size, players, uptime, name
  sortOrder: string; // asc, desc
}

// ============================================================================
// Restore Progress Types
// ============================================================================

export interface RestoreProgress {
  restoreId: number;
  progressPercent: number;
  currentStage: string;
  stageProgress: number;
  totalStages: number;
  elapsedSeconds: number;
  estimatedRemainingSeconds: number;
  bytesProcessed: number;
  bytesTotal: number;
}

export interface RestoreResult {
  restoreId: number;
  success: boolean;
  durationSeconds: number;
  message: string;
  restoredSaveId: number;
  backupCreatedPath?: string;
  warnings: string[];
}

// ============================================================================
// UI Component Types
// ============================================================================

export interface SaveBrowserViewMode {
  type: 'grid' | 'list' | 'timeline' | 'compact';
  columnsPerRow?: number;
  itemsPerPage?: number;
}

export interface SaveBrowserState {
  currentServerid: number;
  saves: AutoSave[];
  selectedSaveIds: number[];
  filters: SaveSearchFilter;
  sortOptions: SaveSortOptions;
  viewMode: SaveBrowserViewMode;
  isLoading: boolean;
  error?: string;
}

export interface RestoreDialogState {
  isOpen: boolean;
  selectedSaveId?: number;
  backupCreated: boolean;
  willCreateBackup: boolean;
  isRestoring: boolean;
  progress?: RestoreProgress;
  error?: string;
}

export interface AutoSaveStore {
  // Save Management
  saves: Map<number, AutoSave>;
  folders: SaveFolder[];
  statistics: SaveStatistics | null;
  healthStatus: SaveHealthStatus | null;
  preferences: AutosavePreferences | null;

  // UI State
  currentServerId: number;
  selectedSaveIds: Set<number>;
  viewMode: SaveBrowserViewMode;
  filters: SaveSearchFilter;
  sortOptions: SaveSortOptions;
  isLoading: boolean;
  error?: string;

  // Restore State
  restoreHistory: SaveRestoreHistory[];
  restorePoints: RestorePoint[];
  restoreInProgress: boolean;

  // Timeline State
  timelineEvents: TimelineEvent[];

  // Actions
  setSaves: (saves: AutoSave[]) => void;
  addSave: (save: AutoSave) => void;
  removeSave: (saveId: number) => void;
  updateSave: (saveId: number, updates: Partial<AutoSave>) => void;
  setFolders: (folders: SaveFolder[]) => void;
  setStatistics: (stats: SaveStatistics | null) => void;
  setHealthStatus: (status: SaveHealthStatus | null) => void;
  setPreferences: (prefs: AutosavePreferences | null) => void;
  setCurrentServerId: (serverId: number) => void;
  toggleSaveSelection: (saveId: number) => void;
  setViewMode: (mode: SaveBrowserViewMode) => void;
  setFilters: (filters: SaveSearchFilter) => void;
  setSortOptions: (options: SaveSortOptions) => void;
  setRestoreHistory: (history: SaveRestoreHistory[]) => void;
  setRestorePoints: (points: RestorePoint[]) => void;
  setTimelineEvents: (events: TimelineEvent[]) => void;
  setLoading: (loading: boolean) => void;
  setError: (error?: string) => void;
  clearSelection: () => void;
}

// ============================================================================
// Drag and Drop Types
// ============================================================================

export interface DragItem {
  type: 'save' | 'save-multiple';
  saveIds: number[];
  sourceFolder?: number;
}

export interface DropItem {
  type: 'folder';
  folderId?: number;
}

File: src-tauri/src/models/autosave.rs
#![allow(dead_code)]
// Auto-Save Management System Models
// Comprehensive data structures for save browsing, restoration, and management

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ============================================================================
// Core Auto-Save Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AutoSave {
    pub id: i64,
    pub server_id: i64,
    pub file_name: String,
    pub file_path: String,
    pub file_size: i64,
    pub checksum: Option<String>,
    pub created_at: String,
    pub save_timestamp: Option<String>,
    pub is_valid: bool,
    pub is_corrupted: bool,
    pub corruption_reason: Option<String>,
    pub player_count: Option<i32>,
    pub uptime_seconds: Option<i32>,
    pub server_version: Option<String>,
    pub mod_count: Option<i32>,
    pub map_name: Option<String>,
    pub is_protected: bool,
    pub custom_label: Option<String>,
    pub notes: Option<String>,
    pub is_favorite: bool,
    pub folder_id: Option<i64>,
    pub created_by: Option<String>,
    pub indexed_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AutoSaveRequest {
    pub server_id: i64,
    pub file_name: String,
    pub file_path: String,
    pub file_size: i64,
    pub player_count: Option<i32>,
    pub uptime_seconds: Option<i32>,
    pub server_version: Option<String>,
    pub mod_count: Option<i32>,
    pub map_name: Option<String>,
}

// ============================================================================
// Save Folder Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveFolder {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub color: String,
    pub parent_folder_id: Option<i64>,
    pub sort_order: i32,
    pub created_at: String,
    pub updated_at: String,
    #[serde(skip)]
    pub children: Vec<SaveFolder>,
    #[serde(skip)]
    pub save_count: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveFolderRequest {
    pub name: String,
    pub description: Option<String>,
    pub color: Option<String>,
    pub parent_folder_id: Option<i64>,
}

// ============================================================================
// Save Restore Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveRestoreHistory {
    pub id: i64,
    pub server_id: i64,
    pub from_save_id: Option<i64>,
    pub to_save_id: i64,
    pub restored_at: String,
    pub restored_by: Option<String>,
    pub restore_duration_seconds: Option<i32>,
    pub restore_method: String, // manual, automatic, scheduled, emergency
    pub success: bool,
    pub error_message: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreRequest {
    pub server_id: i64,
    pub save_id: i64,
    pub create_backup: bool,
    pub restore_method: String,
    pub skip_restart: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveBackupSnapshot {
    pub id: i64,
    pub server_id: i64,
    pub restore_history_id: Option<i64>,
    pub backup_path: String,
    pub backup_size: i64,
    pub created_at: String,
    pub is_valid: bool,
}

// ============================================================================
// Save Metadata Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveMetadata {
    pub id: i64,
    pub auto_save_id: i64,
    pub game_mode: Option<String>,
    pub difficulty_level: Option<String>,
    pub max_players: Option<i32>,
    pub current_players_list: Vec<String>,
    pub mods_list: Vec<ModInfo>,
    pub server_settings: HashMap<String, serde_json::Value>,
    pub world_statistics: HashMap<String, serde_json::Value>,
    pub creatures_count: Option<i32>,
    pub structures_count: Option<i32>,
    pub items_count: Option<i32>,
    pub parsed_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModInfo {
    pub name: String,
    pub version: Option<String>,
    pub mod_id: Option<String>,
}

// ============================================================================
// Validation Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveValidationLog {
    pub id: i64,
    pub auto_save_id: i64,
    pub check_type: String, // integrity, corruption, completeness, compatibility
    pub check_status: String, // passed, failed, warning, skipped
    pub details: HashMap<String, String>,
    pub checked_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveValidationResult {
    pub is_valid: bool,
    pub is_corrupted: bool,
    pub corruption_reason: Option<String>,
    pub warnings: Vec<String>,
    pub errors: Vec<String>,
    pub file_size_ok: bool,
    pub checksum_ok: Option<bool>,
    pub compatible: bool,
}

// ============================================================================
// Restore Points Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestorePoint {
    pub id: i64,
    pub server_id: i64,
    pub auto_save_id: i64,
    pub point_name: String,
    pub description: Option<String>,
    pub point_type: String, // manual, scheduled, critical, post-crash
    pub created_at: String,
    pub is_protected: bool,
    pub created_by: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRestorePointRequest {
    pub server_id: i64,
    pub save_id: i64,
    pub point_name: String,
    pub description: Option<String>,
    pub point_type: Option<String>,
}

// ============================================================================
// Save Comparison Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveComparison {
    pub save_id_1: i64,
    pub save_id_2: i64,
    pub size_difference: i64,
    pub timestamp_difference_seconds: i64,
    pub player_count_difference: i32,
    pub uptime_difference_seconds: i32,
    pub mod_changes: ModChanges,
    pub setting_changes: Vec<SettingChange>,
    pub has_corruption_change: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModChanges {
    pub added_mods: Vec<String>,
    pub removed_mods: Vec<String>,
    pub updated_mods: Vec<ModUpdate>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModUpdate {
    pub mod_name: String,
    pub old_version: String,
    pub new_version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SettingChange {
    pub setting_name: String,
    pub old_value: String,
    pub new_value: String,
}

// ============================================================================
// Timeline Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineEvent {
    pub id: i64,
    pub server_id: i64,
    pub event_type: String, // save_created, restored, protected, deleted, validated
    pub auto_save_id: Option<i64>,
    pub restore_history_id: Option<i64>,
    pub event_time: String,
    pub description: String,
    pub metadata: HashMap<String, String>,
    pub importance_level: String, // low, normal, high, critical
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineResponse {
    pub events: Vec<TimelineEvent>,
    pub total_events: i32,
    pub server_id: i64,
}

// ============================================================================
// Archive and Compression Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveArchive {
    pub id: i64,
    pub server_id: i64,
    pub archive_path: String,
    pub archive_size: i64,
    pub save_ids: Vec<i64>,
    pub created_at: String,
    pub is_verified: bool,
    pub last_accessed: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompressionRequest {
    pub server_id: i64,
    pub save_ids: Vec<i64>,
    pub compression_level: Option<i32>,
}

// ============================================================================
// Cloud Sync Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveCloudSync {
    pub id: i64,
    pub auto_save_id: i64,
    pub cloud_path: Option<String>,
    pub sync_status: String, // pending, syncing, synced, failed
    pub last_sync_at: Option<String>,
    pub sync_error: Option<String>,
    pub cloud_version: Option<String>,
}

// ============================================================================
// Preferences Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AutosavePreferences {
    pub id: i64,
    pub server_id: i64,
    pub auto_index_enabled: bool,
    pub auto_validate_enabled: bool,
    pub auto_compress_old_saves: bool,
    pub compress_after_days: i32,
    pub auto_cleanup_enabled: bool,
    pub cleanup_after_days: i32,
    pub keep_minimum_saves: i32,
    pub create_restore_points: bool,
    pub restore_point_frequency: String, // hourly, daily, weekly, manual
    pub notify_on_restore: bool,
    pub notify_on_corruption: bool,
    pub index_metadata: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdatePreferencesRequest {
    pub server_id: i64,
    pub auto_index_enabled: Option<bool>,
    pub auto_validate_enabled: Option<bool>,
    pub auto_compress_old_saves: Option<bool>,
    pub compress_after_days: Option<i32>,
    pub auto_cleanup_enabled: Option<bool>,
    pub cleanup_after_days: Option<i32>,
    pub keep_minimum_saves: Option<i32>,
    pub create_restore_points: Option<bool>,
    pub restore_point_frequency: Option<String>,
    pub notify_on_restore: Option<bool>,
    pub notify_on_corruption: Option<bool>,
    pub index_metadata: Option<bool>,
}

// ============================================================================
// Statistics and Analytics Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveStatistics {
    pub server_id: i64,
    pub total_saves: i32,
    pub protected_saves: i32,
    pub corrupted_saves: i32,
    pub total_storage_used: i64,
    pub oldest_save_date: Option<String>,
    pub newest_save_date: Option<String>,
    pub average_save_size: i64,
    pub total_restores: i32,
    pub last_restore_date: Option<String>,
    pub favorite_count: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveHealthStatus {
    pub health_score: f64, // 0-100
    pub status: String, // excellent, good, fair, poor, critical
    pub issues_count: i32,
    pub warnings: Vec<String>,
    pub recommendations: Vec<String>,
}

// ============================================================================
// Bulk Operation Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BulkSaveOperation {
    pub operation_type: String, // delete, protect, tag, compress, export
    pub save_ids: Vec<i64>,
    pub operation_data: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BulkOperationResult {
    pub operation_id: i64,
    pub status: String, // pending, in-progress, completed, failed
    pub processed_count: i32,
    pub error_count: i32,
    pub started_at: String,
    pub completed_at: Option<String>,
    pub errors: Vec<String>,
}

// ============================================================================
// Search and Filter Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveSearchFilter {
    pub search_query: Option<String>,
    pub server_id: Option<i64>,
    pub date_range: Option<DateRange>,
    pub status: Option<Vec<String>>, // valid, corrupted, protected, favorite
    pub map_names: Option<Vec<String>>,
    pub min_size: Option<i64>,
    pub max_size: Option<i64>,
    pub player_count_range: Option<(i32, i32)>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DateRange {
    pub start: String,
    pub end: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveSortOptions {
    pub sort_by: String, // date, size, players, uptime, name
    pub sort_order: String, // asc, desc
}

// ============================================================================
// Restore Progress Models
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreProgress {
    pub restore_id: i64,
    pub progress_percent: i32,
    pub current_stage: String,
    pub stage_progress: i32,
    pub total_stages: i32,
    pub elapsed_seconds: i32,
    pub estimated_remaining_seconds: i32,
    pub bytes_processed: i64,
    pub bytes_total: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreResult {
    pub restore_id: i64,
    pub success: bool,
    pub duration_seconds: i32,
    pub message: String,
    pub restored_save_id: i64,
    pub backup_created_path: Option<String>,
    pub warnings: Vec<String>,
}
